lineapy/graph_reader/artifact_collection.py

```diff
@@ -163,39 +169,66 @@ def validate_dependencies(self, dependencies: TaskGraphEdge = {}):
+            edges=self.dependencies,
+        )
+
         # Helper dictionary to look up artifact information by name
-        art_name_to_session_id: Dict[str, LineaID] = {}
+        art_name_to_session_id: Dict[str, str] = {}
         for artifact in self.target_artifacts:
-            art_name_to_session_id[artifact.name] = artifact._session_id
-            art_name_to_session_id[
-                slugify(artifact.name)
-            ] = artifact._session_id
-
-        session_id_nodes = list(self.session_artifacts.keys())
-        session_id_edges = list(
-            chain.from_iterable(
-                (
-                    (
-                        art_name_to_session_id[slugify(from_artname)],
-                        art_name_to_session_id[slugify(to_artname)],
-                    )
-                    for from_artname in from_artname_set
-                )
-                for to_artname, from_artname_set in dependencies.items()
+            art_name_to_session_id[artifact.name] = str(artifact._session_id)
+            art_name_to_session_id[slugify(artifact.name)] = str(
+                artifact._session_id
             )
+
+        inter_session_graph = inter_session_graph.remap_nodes(
+            art_name_to_session_id
         )
+
         # remove loops in graph for dependencies within a session
-        session_id_edges = [
-            edge for edge in session_id_edges if edge[0] != edge[1]
-        ]
-        inter_session_graph = nx.DiGraph()
-        inter_session_graph.add_nodes_from(session_id_nodes)
-        inter_session_graph.add_edges_from(session_id_edges)
+        inter_session_graph.remove_self_loops()
+
         return inter_session_graph
 
+    def create_inter_artifact_taskgraph(self) -> TaskGraph:
+
+        inter_artifact_taskgraph = TaskGraph(nodes=[], edges={})
+
+        # add subgraph for each session_artifact
+        for sa in self.session_artifacts.values():
+            inter_artifact_taskgraph.graph = nx.compose(
+                inter_artifact_taskgraph.graph,
+                sa.nodecollection_dependencies.graph,
+            )
+
+        for (
+            from_session_id,
+            to_session_id,
+        ) in self.inter_session_taskgraph.graph.edges:
+            # get sink nodes from the first session artifact
+            from_session_sink_nodes = self.session_artifacts[
+                from_session_id
+            ].nodecollection_dependencies.sink_nodes
+
+            # get source nodes from the second session artifact
+            to_session_source_nodes = self.session_artifacts[
+                to_session_id
+            ].nodecollection_dependencies.source_nodes
+
+            # add edges between each pair of sink/sources
+
+            inter_artifact_taskgraph.graph.add_edges_from(
+                itertools.product(
+                    from_session_sink_nodes, to_session_source_nodes
+                )
+            )
```

Comment:
This "stitching" of artifacts based on session dependency may interpolate more dependency information than what the user intended/provided. For instance, if a session results in two sink nodes of `A` and `B` and the user provides dependency info from `A` to `C` in a different session, then this current logic will interpolate dependency relation between `B` and `C` as well. Is this something we desire?

In contrast, existing implementation of `combined_taskgraph` directly uses user-specified dependency information to stitch artifacts between sessions. Hence, for the example above, had the user specified dependency between `A` and `C` only, they will not get the interpolated dependency between `B` and `C`.

Reply:
On a related note, is the plan to replace `combined_taskgraph` with `inter_artifact_taskgraph`?

Reply:
This is a great point which is why I wanted a review of this first. 

I wanted to do this stitching because of the following case: 

Session 1 will run `Task A` and `Task B` which creates `Artifact B`. 
Session 2 will run `Task C` and `Task D` which creates `Artifact D`.

In the case that Task C reads from the output of Task B the users cannot specify this dependency easily.

User can only specify dependencies based on Artifacts, say they can only say to run `B >> D` (B before D).

If we don't enforce dependencies on a session level, Task C may run before Task B. 

If you do, you get the problems described above which is osme of the dependencies will be interpolated and may not be correct. 

I personally lean towards over specifying dependencies (bad for performance since less parallelization and bad for readability) instead of potentially having the case above (user literally cannot specify the dependency they want unless they hack our graph)

Reply:
Second point is good, yeah I think `combined_taskgraph` is only used for validating dependencies, which should be done on the `inter_artifact_taskgraph` or `inter_session_taskgraph`. `combined_taskgraph` may not be needed,

Reply:
Interpreting

- `Task C` as common code factored out by LineaPy (e.g., `task_common_var_for_artifact_D_and_downstream`)
- `Task D` as a task responsible for generating `Artifact D` (e.g., `task_D`)

I think the point you made above is valid. That is, `Task B` should be set as a dependency for `Task C`, but relying solely on the user-provided information will miss this, which may result in erroneous parallelization. In fact, the way existing `combined_taskgraph` constructs inter-session dependency seems subject to this error.

As we discussed on a call, the challenge would lie in whether/how we can identify source node(s) related to the target artifact only. That is, had `Task C` above been NOT related to generation of `Artifact D`, then we would ideally NOT create dependency between `Task B` and `Task C` even if `Task C` is a source node of Session 2.

Reply:
I can give a counter example of why we should not create dependency between Task B and Task C. 
Let's say someone is working on the two notebooks(the first notebook has Task A and Task B and the second notebook has Task C and Task D like the same setup in this thread) at the same time(maybe she/he likes two screens/computers).

Task A writes to a side-effect (let's say a.csv) and save artifact a.
Task C reads from the side-effect created by Task A; i.e., read from a.csv, and write to another side-effect (let's say c.csv) save artifact c.
Task B reads from the side-effect created by Task C; i.e., read from c.csv, and save artifact b.

In this case, we should not create dependency between Task B and Task C. 
I'm not saying this is a common scenario, but without tracking input and output side-effects and their versions, I don't see it is possible to add any extra new dependencies that users do not specify.

Reply:
@mingjerli The scenario you outlined above seems to be a case of circular dependency between two sessions, which we are not supporting (error message raised)?